### PR TITLE
Add warn mode for CIP.

### PR DESCRIPTION
**Pull request by vaikas**
Signed-off-by: Ville Aikas <vaikas@chainguard.dev>

#### Summary
Fixes: #98 
Add `mode` field to CIP that allows you to specify 'warn' or 'enforce'. If warn specified the failed policies are returned as webhook warnings and won't block the admission.

#### Release Note
* Fixes: #98 
* Add `mode` field to CIP that allows you to specify 'warn' or 'enforce'. If warn specified the failed policies are returned as webhook warnings and won't block the admission.

#### Documentation
https://github.com/sigstore/sigstore-website/pull/215

### CODE REVIEW

**Comment by vaikas** [comment]
Fulcio was crashlooping again, bumping it up here
https://github.com/sigstore/scaffolding/pull/302

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_types.go [hunk 1]
```diff
@@ -61,6 +61,12 @@ type ClusterImagePolicySpec struct {
 	// not even get evaluated, as the Policy is considered failed.
 	// +optional
 	Policy *Policy `json:"policy,omitempty"`
+	// Mode controls whether a failing policy will be rejected (not admitted),
+	// or if errors are converted to Warnings.
+	// enforce - Reject (default)
+	// warn - allow but warn
+	// +optional
```
Comment (JAORMX) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_types.go:68]: If kubebuilder is used, you could probably also default it here:
```
+kubebuilder:default:=enforce
```

That would trickle down to a similar setting in the CRD which would be quite handy.

**Review by JAORMX — CHANGES_REQUESTED** [review]

**Comment by JAORMX** [comment]
@vaikas could it be related to the kubectl version you're using?

**Comment by vaikas** [comment]
Ok, so the root cause is here:
https://github.com/knative/pkg/issues/2581

TL;DR: anything with a `\n` gets silently dropped on the floor... Whee...

Note that this also means that https://github.com/sigstore/policy-controller/pull/125 warnings don't make it to `kubectl` either.

**Comment by vaikas** [comment]
Ok, well, that was a fun detour. Now, I can create a 'warn' policy that fails the signature checks.

```
vaikas@villes-mbp pkg % kubectl create -f ~/demo/pod.yaml
Warning: failed policy: image-policy-keyless: spec.containers[0].image
Warning: registry.local:5000/sigstore/demo@sha256:a420f0e3a1f8c20c6be3caaca3d6601139806b7dc5d7e4256a912675cfd319c6 signature keyless validation failed for authority authority-0 for registry.local:5000/sigstore/demo@sha256:a420f0e3a1f8c20c6be3caaca3d6601139806b7dc5d7e4256a912675cfd319c6: Get "https://registry.local:5000/v2/": dial tcp: lookup registry.local on 10.12.0.10:53: no such host; Get "http://registry.local:5000/v2/": dial tcp: lookup registry.local on 10.12.0.10:53: no such host
pod/testpod created
```

**Comment by vaikas** [comment]
Well, hippity hop, looks like it's having trouble pulling the image from ghcr.io, or at least I take it that it's taking over 100 seconds to pull this image:
```
  Events:
    Type     Reason       Age                   From               Message
    ----     ------       ----                  ----               -------
    Normal   Scheduled    2m11s                 default-scheduler  Successfully assigned ctlog-system/ctlog-00001-deployment-6689d95b74-jszt9 to kind-worker
    Warning  FailedMount  2m8s (x4 over 2m11s)  kubelet            MountVolume.SetUp failed for volume "config" : configmap references non-existent config key: config
    Warning  FailedMount  2m8s (x4 over 2m11s)  kubelet            MountVolume.SetUp failed for volume "keys" : secret "ctlog-secret" not found
    Normal   Pulling      2m3s                  kubelet            Pulling image "ghcr.io/sigstore/scaffolding/ct_server@sha256:6d424eb63c61af2beb80a80b2c6804fc105dcd8dd83f37d49f8618fa930f2a87"

```

**Comment by vaikas** [comment]
Compared to something like this:
```
    Normal   Pulling      2m52s              kubelet            Pulling image "gcr.io/projectsigstore/fulcio@sha256:61081295a8f75ed7537b5d1f8c7320e078dc00e4562c0bf605fbefa062c690de"
    Normal   Pulled       2m49s              kubelet            Successfully pulled image "gcr.io/projectsigstore/fulcio@sha256:61081295a8f75ed7537b5d1f8c7320e078dc00e4562c0bf605fbefa062c690de" in 2.155591735s

```

**Review by JAORMX — APPROVED** [review]
LGTM

File: cmd/tester/main.go [hunk 2]
```diff
@@ -126,12 +129,22 @@ func main() {
 
 	result, errs := webhook.ValidatePolicy(ctx, ns, ref, *cip, remoteOpts...)
 	errStrings := []string{}
+	warningStrings := []string{}
 	for _, err := range errs {
+		var fe *apis.FieldError
+		if errors.As(err, &fe) {
+			if fe.Level == apis.WarningLevel {
```
Comment (mattmoor) [inline cmd/tester/main.go:136]: `apis.FieldError` can contain a mix of errors and warnings.

We want to partition the errors and warnings with this filter method: https://github.com/knative/pkg/blob/a6afcab5bd78047384d858c913826af80dd7dcd8/apis/field_error.go#L168-L172

So `fe.Filter(apis.WarningLevel)` goes into `warningStrings`, and `fe.Filter(apis.ErrorLevel)` goes into `errStrings`

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_defaults_test.go [hunk 3]
```diff
@@ -52,6 +52,34 @@ func TestNameDefaulting(t *testing.T) {
...
 }
 
+func TestModeDefaulting(t *testing.T) {
+	tests := []struct {
+		name     string
+		mode     string
+		wantMode string
+	}{{name: "empty",
+		wantMode: "enforce",
+	}, {
```
Comment (mattmoor) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_defaults_test.go:62]: nit: fix spacing

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go [hunk 4]
```diff
@@ -39,6 +39,9 @@ var (
 	// TODO(vaikas): Consider adding a warn which would pass but use
 	// `warn` as return type for the webhook response.
```
Comment (mattmoor) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:40]: Isn't this covered by:
```yaml
static: { action: fail }
mode: warn
```

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go [hunk 5]
```diff
@@ -226,7 +231,7 @@ func (p *Policy) Validate(ctx context.Context) *apis.FieldError {
 func (identity *Identity) Validate(ctx context.Context) *apis.FieldError {
 	var errs *apis.FieldError
 	if identity.Issuer == "" && identity.Subject == "" && identity.IssuerRegExp == "" && identity.SubjectRegExp == "" {
-		errs = errs.Also(apis.ErrMissingField("issuer", "subject", "issuerRegExp", "subjectRegExp"))
+		return apis.ErrMissingField("issuer", "subject", "issuerRegExp", "subjectRegExp")
```
Comment (mattmoor) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:234]: We should drop this check entirely in favor of your checks below.  This is the breaking change I was hoping to warn folks about first IIUC.

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go [hunk 6]
```diff
@@ -240,6 +245,12 @@ func (identity *Identity) Validate(ctx context.Context) *apis.FieldError {
 	if identity.SubjectRegExp != "" {
 		errs = errs.Also(ValidateRegex(identity.SubjectRegExp).ViaField("subjectRegExp"))
 	}
+	if identity.SubjectRegExp == "" && identity.Subject == "" {
+		errs = errs.Also(apis.ErrMissingField("subject", "subjectRegExp").At(apis.WarningLevel))
+	}
+	if identity.IssuerRegExp == "" && identity.Issuer == "" {
+		errs = errs.Also(apis.ErrMissingField("issuer", "issuerRegExp").At(apis.WarningLevel))
+	}
```
Comment (mattmoor) [inline pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:253]: I think that we may actually want these to be errors.  If we see an identity without one of these, then it is an error.

I think what should be a warning is an empty list of Identities: https://github.com/sigstore/policy-controller/blob/6a0a54404d9c341e70faf98fd1b0f7293efd64e4/pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go#L175-L177

I don't see that check.

File: pkg/apis/policy/v1beta1/clusterimagepolicy_defaults_test.go [hunk 7]
```diff
@@ -52,6 +52,34 @@ func TestNameDefaulting(t *testing.T) {
 	}
 }
 
+func TestModeDefaulting(t *testing.T) {
+	tests := []struct {
+		name     string
+		mode     string
+		wantMode string
+	}{{name: "empty",
+		wantMode: "enforce",
```
Comment (mattmoor) [inline pkg/apis/policy/v1beta1/clusterimagepolicy_defaults_test.go:61]: same nit

File: pkg/apis/policy/v1beta1/clusterimagepolicy_validation.go [hunk 8]
```diff
@@ -39,6 +39,9 @@ var (
 	// TODO(vaikas): Consider adding a warn which would pass but use
 	// `warn` as return type for the webhook response.
```
Comment (mattmoor) [inline pkg/apis/policy/v1beta1/clusterimagepolicy_validation.go:40]: same

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 4 above)
Comment (vaikas) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:40]: Yes, forgot to remove the todo :)

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 6 above)
Comment (vaikas) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:253]: Well, we did not require either to be there. So having subject, but no issuer was valid choice so I don't think they should be errors.
I think having empty identities as a warning makes sense for sure though.

**Comment by vaikas** [comment]
I'll tackle this as a follow up to this PR:
https://github.com/sigstore/policy-controller/pull/125#pullrequestreview-1069638971

File: cmd/tester/main.go [hunk 9]
```diff
@@ -126,12 +129,25 @@ func main() {
...
 	errStrings := []string{}
+	warningStrings := []string{}
 	for _, err := range errs {
-		errStrings = append(errStrings, strings.Trim(err.Error(), "\n"))
+		var fe *apis.FieldError
+		if errors.As(err, &fe) {
+			if warnFE := fe.Filter(apis.WarningLevel); warnFE != nil {
+				warningStrings = append(warningStrings, strings.Trim(warnFE.Error(), "\n"))
+			}
+			if errorFE := fe.Filter(apis.WarningLevel); errorFE != nil {
```
Comment (mattmoor) [inline cmd/tester/main.go:139]:
```suggestion
			if errorFE := fe.Filter(apis.ErrorLevel); errorFE != nil {
```

File: cmd/tester/main.go [hunk 10]
```diff
@@ -126,12 +129,25 @@ func main() {
...
+				warningStrings = append(warningStrings, strings.Trim(errorFE.Error(), "\n"))
```
Comment (mattmoor) [inline cmd/tester/main.go:140]:
```suggestion
				errStrings = append(errStrings, strings.Trim(errorFE.Error(), "\n"))
```

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 6 above)
Comment (mattmoor) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:252]: For these, do you want to indicate to folks that these will become required in a future release as part of the error message?

**Review by mattmoor — APPROVED** [review]

File: pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go (hunk 6 above)
Comment (vaikas) [reply pkg/apis/policy/v1alpha1/clusterimagepolicy_validation.go:252]: Yeah, but I'll do that as part of the work below for proper e2e tests that weren't there. Want to figure out when we start enforcing it :)

**Review by dlorenc — APPROVED** [review]
